File: src/squares/js/squares.js
/* @flow */

export type Board = {|
  +hMatrix: Array<Array<boolean>>, // row-major, size x (size - 1)
  +size: number,
  +vMatrix: Array<Array<boolean>>, // row-major, (size - 1) x size
|};

export default function squares(board: Board): Map<number, number> {
  const map = new Map();
  // i represents the size of the square we are checking. We will check for the
  // existence of squares of size i, then continue iterating to larger sizes
  // until we reach the max size of any square that can exist on the board.
  for (let i = 1; i < board.size; ++i) {
    let count = 0;

    // j represents the starting row we want to check for the squares. We will
    // check for squares of size i that start on row j.
    for (let j = 0; j < board.size - i; ++j) {
      // k represents the starting column we want to check for the squares.
      // Given the size of the squares (i) and the starting row of the
      // squares (j), we need to pick the column that we are inspecting for
      // squares (k).
      for (let k = 0; k < board.size - i; ++k) {
        // In this loop we are testing for the existence of a specific square.
        // We need to query the hMatrix and vMatrix to see if this square
        // exists.
        if (_checkIfSquareExists(board, i, j, k)) {
          count += 1;
        }
      }
    }
    map.set(i, count);
  }

  return map;
}

function _checkIfSquareExists(
  board: Board,
  size: number,
  row: number,
  col: number,
): boolean {
  const { hMatrix, vMatrix } = board;

  // Check the top and bottom rows for all the necessary horizontal lines.
  for (let i = row; i < row + size; ++i) {
    if (!vMatrix[i][col] || !vMatrix[i][col + size]) {
      return false;
    }
  }

  // Check the first and last columns for all the necessary vertical lines.
  for (let i = col; i < col + size; ++i) {
    if (!hMatrix[row][i] || !hMatrix[row + size][i]) {
      return false;
    }
  }

  return true;
}

File: src/squares/py/squares.py

def squares(board):
    size, h_matrix, v_matrix = board

    count_dict = {}

    # We want to enumerate all possible sizes that a square can be within
    # a dot matrix of (size)-by-(size) dimensions. i represents the size of the
    # squares we are currently checking.
    for i in range(1, size):
        count = 0
        # For each size i, we need to test all possible locations of a square.
        # j represents the row of a square we are checking.
        for j in range(size - i):
            # For each size i, we need to test all possible locations of a
            # square. k represents the column of a square we are checking.
            for k in range(size - i):
                if does_square_exist(board, square_size=i, row=j, col=k):
                    count += 1

        count_dict[i] = count

    return count_dict


def does_square_exist(board, square_size, row, col):
    _, h_matrix, v_matrix = board

    # Check if the vertical edges of the square exist.
    for i in range(row, row + square_size):
        if not v_matrix[i][col] or not v_matrix[i][col + square_size]:
            return False

    # Check if the horizontal edges of the square exist.
    for i in range(col, col + square_size):
        if not h_matrix[row][i] or not h_matrix[row + square_size][i]:
            return False

    return True
